WcaOnRails/app/assets/javascripts/registrations.js

```diff
@@ -36,3 +36,21 @@ onPage('registrations#index', function() {
   // This makes the interface resemble the order.
   $('.name .sortable').addClass('asc');
 });
+
+function comparePaymentDate(a, b) {
+  var elemA = $(a);
+  var elemB = $(b);
+  var paidA = elemA.data("paidDate");
+  var paidB = elemB.data("paidDate");
+  if (paidA !== "" && paidB !== "") {
+    // Both have paid, compare their last payment dates
+    return paidA.localeCompare(paidB);
+  } else if (paidA === "" && paidB === "") {
+    // None have paid, compare their registration dates
+    return elemA.data("registeredAt").localeCompare(elemB.data("registeredAt"));
```

Comment:
It doesn't violate transitivity: in your scenario the 3 people sorting is actually Philippe < Lars < Jeremy!

- compare(Philippe, Jeremy) gives -1 (goes into the last branch)
- compare(Jeremy, Philippe) gives 1 (goes into the 3rd branch)
- compare(Jeremy, Lars) gives 1 (goes into the 3rd branch)
- compare(Lars, Jeremy) gives -1 (goes into the last branch)
- compare(Philippe, Lars) gives -1 (first branch, paidPhilippe.compare(paidLars) = -1)
- compare(Lars, Philippe) gives 1 (first branch, paidLars.compare(paidPhilippe) = 1)

In actual sentences, the whole logic does the following:
- If both people have paid, lowest paid date first (lowest meaning earliest in this case).
- If both people have not paid, lowest registration date first.
- Else the one who has paid is lower than the other.